File: barcode-QRcodeScanner/barcode-QRcodeScanner.cpp
#include <algorithm>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/objdetect.hpp>
#include <opencv2/objdetect/barcode.hpp>

namespace fs = std::filesystem;

struct CodeResult {
    std::string type;
    std::string data;
    std::vector<cv::Point2f> points;
};

std::vector<std::vector<cv::Point2f>> polygonsFromMat(const cv::Mat& points) {
    std::vector<std::vector<cv::Point2f>> polygons;
    if (points.empty()) {
        return polygons;
    }
    if (points.type() != CV_32FC2 || points.cols != 4) {
        throw std::runtime_error("OpenCV returned an unexpected polygon layout");
    }

    polygons.reserve(points.rows);
    for (int row = 0; row < points.rows; ++row) {
        const cv::Point2f* begin = points.ptr<cv::Point2f>(row);
        polygons.emplace_back(begin, begin + points.cols);
    }
    return polygons;
}

std::vector<CodeResult> decodeQrCodes(const cv::Mat& image) {
    cv::QRCodeDetector detector;
    std::vector<std::string> decoded_info;
    cv::Mat points;
    std::vector<CodeResult> results;

    if (detector.detectAndDecodeMulti(image, decoded_info, points)) {
        const auto polygons = polygonsFromMat(points);
        const std::size_t count = std::min(decoded_info.size(), polygons.size());
        for (std::size_t index = 0; index < count; ++index) {
            if (!decoded_info[index].empty()) {
                results.push_back(
                    {"QR_CODE", decoded_info[index], polygons[index]});
            }
        }
    }
    if (!results.empty()) {
        return results;
    }

    cv::Mat single_points;
    const std::string data = detector.detectAndDecode(image, single_points);
    const auto polygons = polygonsFromMat(single_points);
    if (!data.empty() && !polygons.empty()) {
        results.push_back({"QR_CODE", data, polygons.front()});
    }
    return results;
}

std::vector<CodeResult> decodeBarcodes(const cv::Mat& image) {
    cv::barcode::BarcodeDetector detector;
    std::vector<std::string> decoded_info;
    std::vector<std::string> decoded_types;
    cv::Mat points;
    std::vector<CodeResult> results;

    if (!detector.detectAndDecodeWithType(
            image, decoded_info, decoded_types, points)) {
        return results;
    }

    const auto polygons = polygonsFromMat(points);
    const std::size_t count =
        std::min({decoded_info.size(), decoded_types.size(), polygons.size()});
    for (std::size_t index = 0; index < count; ++index) {
        if (!decoded_info[index].empty()) {
            results.push_back(
                {decoded_types[index], decoded_info[index], polygons[index]});
        }
    }
    return results;
}

cv::Mat annotateCodes(
    const cv::Mat& image, const std::vector<CodeResult>& results) {
    cv::Mat annotated = image.clone();
    for (const CodeResult& result : results) {
        std::vector<cv::Point> polygon;
        polygon.reserve(result.points.size());
        for (const cv::Point2f& point : result.points) {
            polygon.emplace_back(cvRound(point.x), cvRound(point.y));
        }
        cv::polylines(annotated, polygon, true, cv::Scalar(30, 190, 30), 3,
                      cv::LINE_AA);

        const auto leftmost = std::min_element(
            polygon.begin(), polygon.end(),
            [](const cv::Point& left, const cv::Point& right) {
                return left.x < right.x;
            });
        const auto topmost = std::min_element(
            polygon.begin(), polygon.end(),
            [](const cv::Point& left, const cv::Point& right) {
                return left.y < right.y;
            });
        const cv::Point anchor(
            leftmost == polygon.end() ? 0 : leftmost->x,
            std::max(24, (topmost == polygon.end() ? 24 : topmost->y - 10)));
        cv::putText(
            annotated, result.type + ": " + result.data, anchor,
            cv::FONT_HERSHEY_SIMPLEX, 0.55, cv::Scalar(25, 25, 220), 2,
            cv::LINE_AA);
    }
    return annotated;
}

void printUsage(const char* program) {
    std::cerr << "Usage: " << program
              << " [image] [--output FILE] [--no-qr] [--no-barcode] "
                 "[--display]\n";
}

int main(int argc, char** argv) {
    fs::path image_path = "zbar-test.jpg";
    fs::path output_path = "output/decoded-codes.png";
    bool scan_qr = true;
    bool scan_barcodes = true;
    bool display = false;
    bool positional_image_seen = false;

    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        if (argument == "--output" && index + 1 < argc) {
            output_path = argv[++index];
        } else if (argument == "--no-qr") {
            scan_qr = false;
        } else if (argument == "--no-barcode") {
            scan_barcodes = false;
        } else if (argument == "--display") {
            display = true;
        } else if (argument == "--help" || argument == "-h") {
            printUsage(argv[0]);
            return 0;
        } else if (!argument.empty() && argument.front() != '-' &&
                   !positional_image_seen) {
            image_path = argument;
            positional_image_seen = true;
        } else {
            printUsage(argv[0]);
            return 2;
        }
    }

    if (!scan_qr && !scan_barcodes) {
        std::cerr << "error: both decoders are disabled\n";
        return 2;
    }

    try {
        const cv::Mat image = cv::imread(image_path.string(), cv::IMREAD_COLOR);
        if (image.empty()) {
            throw std::runtime_error("Could not read input image: " +
                                     image_path.string());
        }

        std::vector<CodeResult> results;
        if (scan_qr) {
            auto qr_results = decodeQrCodes(image);
            results.insert(
                results.end(), qr_results.begin(), qr_results.end());
        }
        if (scan_barcodes) {
            auto barcode_results = decodeBarcodes(image);
            results.insert(
                results.end(), barcode_results.begin(), barcode_results.end());
        }

        for (const CodeResult& result : results) {
            std::cout << result.type << '\t' << result.data << '\n';
        }
        std::cout << "decoded_count=" << results.size() << '\n';

        const cv::Mat annotated = annotateCodes(image, results);
        if (output_path.has_parent_path()) {
            fs::create_directories(output_path.parent_path());
        }
        if (!cv::imwrite(output_path.string(), annotated)) {
            throw std::runtime_error(
                "OpenCV could not write output image: " +
                output_path.string());
        }
        std::cout << "output=" << fs::absolute(output_path) << '\n';

        if (display) {
            cv::imshow("Decoded codes", annotated);
            cv::waitKey(0);
            cv::destroyAllWindows();
        }
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 2;
    }
    return 0;
}

File: barcode-QRcodeScanner/barcode-QRcodeScanner.py
#!/usr/bin/env python3
"""Decode QR codes and supported retail barcodes with OpenCV."""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

import cv2
import numpy as np
from numpy.typing import NDArray

Image = NDArray[np.uint8]


@dataclass(frozen=True)
class CodeResult:
    code_type: str
    data: str
    points: NDArray[np.float32]


def _normalise_polygons(points: object) -> list[NDArray[np.float32]]:
    if points is None:
        return []
    array = np.asarray(points, dtype=np.float32)
    if array.size == 0:
        return []
    return [polygon.reshape(-1, 2) for polygon in array.reshape(-1, 4, 2)]


def decode_qr_codes(image: Image) -> list[CodeResult]:
    """Decode every QR code OpenCV can resolve in an image."""
    if image is None or image.size == 0:
        raise ValueError("decode_qr_codes expects a non-empty image")

    detector = cv2.QRCodeDetector()
    detected, decoded_info, points, _ = detector.detectAndDecodeMulti(image)
    polygons = _normalise_polygons(points)
    results = [
        CodeResult("QR_CODE", data, polygon)
        for data, polygon in zip(decoded_info if detected else (), polygons)
        if data
    ]
    if results:
        return results

    # Some OpenCV builds detect a single code more reliably through the
    # single-code path, so retain this deterministic fallback.
    data, single_points, _ = detector.detectAndDecode(image)
    single_polygons = _normalise_polygons(single_points)
    if data and single_polygons:
        return [CodeResult("QR_CODE", data, single_polygons[0])]
    return []


def _create_barcode_detector() -> object:
    namespace = getattr(cv2, "barcode", None)
    if namespace is not None and hasattr(namespace, "BarcodeDetector"):
        return namespace.BarcodeDetector()
    legacy_alias = getattr(cv2, "barcode_BarcodeDetector", None)
    if legacy_alias is not None:
        return legacy_alias()
    raise RuntimeError(
        "This OpenCV build has no BarcodeDetector. Install opencv-python "
        "4.8+ or an opencv-contrib-python build that includes barcode."
    )


def decode_barcodes(image: Image) -> list[CodeResult]:
    """Decode EAN-8, EAN-13, UPC-A, and UPC-E barcodes supported by OpenCV."""
    if image is None or image.size == 0:
        raise ValueError("decode_barcodes expects a non-empty image")

    detector = _create_barcode_detector()
    detected, decoded_info, decoded_types, points = (
        detector.detectAndDecodeWithType(image)
    )
    polygons = _normalise_polygons(points)
    if not detected:
        return []
    return [
        CodeResult(code_type, data, polygon)
        for data, code_type, polygon in zip(
            decoded_info, decoded_types, polygons
        )
        if data
    ]


def annotate_codes(image: Image, results: Iterable[CodeResult]) -> Image:
    """Draw decoded quadrangles and readable labels on a copy of the image."""
    annotated = image.copy()
    for result in results:
        polygon = np.rint(result.points).astype(np.int32).reshape(-1, 1, 2)
        cv2.polylines(annotated, [polygon], True, (30, 190, 30), 3, cv2.LINE_AA)
        anchor_x = int(np.min(polygon[:, 0, 0]))
        anchor_y = max(24, int(np.min(polygon[:, 0, 1])) - 10)
        label = f"{result.code_type}: {result.data}"
        cv2.putText(
            annotated,
            label,
            (anchor_x, anchor_y),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.55,
            (25, 25, 220),
            2,
            cv2.LINE_AA,
        )
    return annotated


def scan_image(
    image: Image, *, scan_qr: bool = True, scan_barcodes: bool = True
) -> list[CodeResult]:
    results: list[CodeResult] = []
    if scan_qr:
        results.extend(decode_qr_codes(image))
    if scan_barcodes:
        results.extend(decode_barcodes(image))
    return results


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Decode QR codes plus EAN/UPC retail barcodes using native "
            "OpenCV detectors."
        )
    )
    parser.add_argument(
        "image",
        nargs="?",
        type=Path,
        default=Path(__file__).with_name("zbar-test.jpg"),
        help="Input image (default: bundled zbar-test.jpg)",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("output/decoded-codes.png"),
        help="Annotated output image",
    )
    parser.add_argument("--no-qr", action="store_true", help="Skip QR decoding")
    parser.add_argument(
        "--no-barcode", action="store_true", help="Skip EAN/UPC barcode decoding"
    )
    parser.add_argument("--display", action="store_true", help="Open a GUI window")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if args.no_qr and args.no_barcode:
        print("error: both decoders are disabled")
        return 2

    image = cv2.imread(str(args.image), cv2.IMREAD_COLOR)
    if image is None:
        print(f"error: could not read input image: {args.image}")
        return 2

    try:
        results = scan_image(
            image, scan_qr=not args.no_qr, scan_barcodes=not args.no_barcode
        )
    except (RuntimeError, ValueError) as error:
        print(f"error: {error}")
        return 2

    for result in results:
        print(f"{result.code_type}\t{result.data}")
    print(f"decoded_count={len(results)}")

    try:
        annotated = annotate_codes(image, results)
        args.output.parent.mkdir(parents=True, exist_ok=True)
        if not cv2.imwrite(str(args.output), annotated):
            raise OSError(f"OpenCV could not write output image: {args.output}")
    except (OSError, cv2.error) as error:
        print(f"error: {error}")
        return 2
    print(f"output={args.output.resolve()}")

    if args.display:
        cv2.imshow("Decoded codes", annotated)
        cv2.waitKey(0)
        cv2.destroyAllWindows()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
